### PR TITLE
compiler: Improve aliases detection, processing, and optimization

### LINKED ISSUE #1627: un-floatified divisions
Reproducer:

```
    grid = Grid(shape=(10, 10)) 
    x, y = grid.dimensions

    u = TimeFunction(name='u', grid=grid)

    eq = Eq(u.forward, ((u/x.spacing) + 2.0)/x.spacing)

    op = Operator(eq)

    exprs = FindNodes(Expression).visit(op)
    assert len(exprs) == 2 
    assert ccode(exprs[0].expr.rhs) == "1.0F/h_x"
```

it's generating `1.0/h_x` while it should generate `1.0F/h_x`

### PR DESCRIPTION
**Pull request by FabioLuporini**
This PR ships significant improvements to the CIRE passes:

* several critical code paths rewritten
* generalizes hoisting of invariant expressions to arbitrary dimensions, and in particular drop the hacky `make_is_time_invariant` machinery
* improves the cost model, used to decide what aliases deserve a tensor temporary and what not
* more tests
* fix previously x-failing tests
* removes now irrelevant knobs that were exposed to the user to control CIRE (e.g., `cire-repeats-sops`). 

fixes #1627

### CODE REVIEW

File: devito/symbolics/manipulation.py [hunk 1]
```diff
@@ -27,20 +28,31 @@ def uxreplace(expr, rule):
 
     By avoiding re-evaluations, this function is typically much quicker than
     SymPy's xreplace.
+
+    A further feature of ``uxreplace`` consists of enabling the substitution
+    of compound nodes. Consider the expression `a*b*c*d`; if one wants to replace
+    `b*c` with say `e*f`, then the following mapper may be passed:
+    `{a*b*c*d: {b: e*f, c: None}}`. This way, only the `b` and `c` pertaining to
```
Comment (georgebisbas) [inline devito/symbolics/manipulation.py:35]: So through this way, replacing also holds if b and c do not appear in sequence...(like the one after the other) ?

**Review by georgebisbas — APPROVED** [review]

File: devito/symbolics/manipulation.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/symbolics/manipulation.py:35]: correct

File: devito/passes/clusters/aliases.py [hunk 2]
```diff
@@ -818,12 +831,40 @@ def lower_schedule(cluster, schedule, chosen, sregistry, options):
...
+            best = i
+            continue
+
+        i_flop_score, i_ws_score = i.score
+
+        # The current heustic is fairly basic: the one with smaller working
+        # set size increase wins, unless there's a massive reduction in operation
+        # count in the other one
+        delta = i_ws_score - best_ws_score
+        if (delta > 0 and i_flop_score / best_flop_score > 100) or \
```
Comment (mloubout) [inline devito/passes/clusters/aliases.py:852]: 100? Shouldn't it bit some MIN_COST too?

File: devito/passes/clusters/aliases.py [hunk 3]
```diff
@@ -1130,3 +1166,35 @@ def wset(exprs):
...
+def potential_max_deriv_order(exprs):
+    """
+    The maximum FD derivative order in a list of expressions.
+    """
+    # NOTE: e might propagate the Derivative(...) information down from the
+    # symbolic language, but users may do crazy things and write their own custom
+    # expansions "by hand" (i.e., not resorting to Derivative(...)), hence instead
+    # of looking for Derivative(...) we use the following heuristic:
+    #   add(mul, mul, ...) -> stems from first order derivative
+    #   add(mul(add(mul, mul, ...), ...), ...) -> stems from second order derivative
```
Comment (mloubout) [inline devito/passes/clusters/aliases.py:1180]: Not completely convinced by this heuristic eventhough I think it makes sense but may miss some cases (i.e pure `Derivative(Derivative(expr))`)

File: devito/symbolics/manipulation.py [hunk 4]
```diff
@@ -94,6 +107,53 @@ def _eval_numbers(expr, args):
...
+class Uxmapper(dict):
+
+    """
+    A helper mapper for uxreplace. Any dict can be passed as subs to uxreplace,
+    but with a Uxmapper it is also possible to easily express compound replacements,
+    such as sub-expressions within n-ary operations (e.g., `a*b` inside the 4-way
+    Mul `a*c*b*d`).
+    """
+
+    class Uxsubmap(dict):
```
Comment (mloubout) [inline devito/symbolics/manipulation.py:119]: Isn't it a bit nicer without decalring that class within a class

**Review by mloubout — COMMENTED** [review]
Minor comments but looks good to me.

File: devito/passes/clusters/aliases.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/passes/clusters/aliases.py:852]: I have a subsequent (large) PR which changes lots of things around here, so I'll definitely do that but can it wait until the next PR?

File: devito/symbolics/manipulation.py (hunk 4 above)
Comment (FabioLuporini) [reply devito/symbolics/manipulation.py:119]: not sure, its purpose is limited to the internals of Uxmapper (hence the name)

Comment (mloubout) [reply devito/symbolics/manipulation.py:119]: ok

File: devito/passes/clusters/aliases.py (hunk 2 above)
Comment (mloubout) [reply devito/passes/clusters/aliases.py:852]: Fine with me

**Review by mloubout — APPROVED** [review]

**Pull request merged**